### ISSUE TITLE
CLANG and METAL backends fail with bfloat16 when ran with Clang v14

### ISSUE THREAD

**Issue by skoshx** [tinygrad/tinygrad#3453]
Installed tinygrad from scratch, then just ran:

```
$ CLANG=1 python3 examples/coder.py
```
and
```
$ METAL=1 python3 examples/coder.py
```

Here's debug output (quite same on METAL and CLANG, both contain `__bf16*`)
```c
create model:  32.15 ms
download weights:  25.70 ms
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%| | assign <LB CLANG (4096, 4096) contig:True (<BinaryOps.ADD: 1>, None)> <- <LB CLANG (4096, 4096) contig:True (<UnaryOps.CAST: 3>, None)>
*** DISK:/U    1 empty 9943044915 dtypes.uchar          arg   1 mem  0.00 GB 
*** CLANG      2 read   33.55M,   CLANG <- DISK:/U      arg   2 mem  0.03 GB tm   2508.54us/     2.51ms (    0.00 GFLOPS,   13.38 GB/s)
  0 ━┳ STORE MemBuffer(idx=0, dtype=dtypes.half, st=ShapeTracker(views=(View(shape=(4096, 4096), strides=(4096, 1), offset=0, mask=None, contiguous=True),)))
  1  ┗━┳ CAST (dtypes.half, False)
  2    ┗━┳ CAST (dtypes.float, True)
  3      ┗━┳ MUL 
  4        ┣━┳ CAST (dtypes.uint, False)
  5        ┃ ┗━┳ CAST (dtypes.ushort, True)
  6        ┃   ┗━━ LOAD MemBuffer(idx=1, dtype=dtypes.bfloat16, st=ShapeTracker(views=(View(shape=(32, 64, 2, 4096), strides=(524288, 4096, 262144, 1), offset=0, mask=None, contiguous=False), View(shape=(4096, 4096), strides=(4096, 1), offset=0, mask=None, contiguous=True))))
  7        ┗━━ CONST ConstBuffer(val=65536, dtype=dtypes.uint, st=ShapeTracker(views=(View(shape=(4096, 4096), strides=(0, 0), offset=0, mask=None, contiguous=False),)))
float4 merging axis : [(2, 2, 1, 4), (2, 8192, 0, 4)]
void E_4096_1024_4(half* restrict data0, const __bf16* restrict data1) {
  for (int ridx0 = 0; ridx0 < 4096; ridx0++) {
    for (int ridx1 = 0; ridx1 < 1024; ridx1++) {
      int alu0 = (ridx1*4);
      int alu1 = (((ridx0/128)*524288)+(((ridx0/2)%64)*4096)+((ridx0%2)*262144)+alu0);
      __bf16 val0 = data1[alu1];
      __bf16 val1 = data1[alu1+1];
      __bf16 val2 = data1[alu1+2];
      __bf16 val3 = data1[alu1+3];
      __bf16 precast0 = val0;
      __bf16 precast1 = val1;
      __bf16 precast2 = val2;
      __bf16 precast3 = val3;
      unsigned int precast4 = ((unsigned int)((*((unsigned short*)&precast0)))*(unsigned int)(65536));
      unsigned int precast5 = ((unsigned int)((*((unsigned short*)&precast1)))*(unsigned int)(65536));
      unsigned int precast6 = ((unsigned int)((*((unsigned short*)&precast2)))*(unsigned int)(65536));
      unsigned int precast7 = ((unsigned int)((*((unsigned short*)&precast3)))*(unsigned int)(65536));
      int alu2 = ((ridx0*4096)+alu0);
      data0[alu2] = (half)((*((float*)&precast4)));
      data0[alu2+1] = (half)((*((float*)&precast5)));
      data0[alu2+2] = (half)((*((float*)&precast6)));
      data0[alu2+3] = (half)((*((float*)&precast7)));
    }
  }
}
<stdin>:7:48: error: __bf16 is not supported on this target
void E_4096_1024_4(half* restrict data0, const __bf16* restrict data1) {
                                               ^
<stdin>:12:7: error: __bf16 is not supported on this target
      __bf16 val0 = data1[alu1];
      ^
<stdin>:13:7: error: __bf16 is not supported on this target
      __bf16 val1 = data1[alu1+1];
      ^
<stdin>:14:7: error: __bf16 is not supported on this target
      __bf16 val2 = data1[alu1+2];
      ^
<stdin>:15:7: error: __bf16 is not supported on this target
      __bf16 val3 = data1[alu1+3];
      ^
<stdin>:16:7: error: __bf16 is not supported on this target
      __bf16 precast0 = val0;
      ^
<stdin>:17:7: error: __bf16 is not supported on this target
      __bf16 precast1 = val1;
      ^
<stdin>:18:7: error: __bf16 is not supported on this target
      __bf16 precast2 = val2;
      ^
<stdin>:19:7: error: __bf16 is not supported on this target
      __bf16 precast3 = val3;
      ^
9 errors generated.
ram used:  0.00 GB, layers.0.attention.wq.weight                      :   0%| | 
loaded weights in  49.11 ms, 0.03 GB loaded at 0.68 GB/s
weights -> model:  67.75 ms
Traceback (most recent call last):
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/examples/coder.py", line 40, in <module>
    nn.state.load_state_dict(model, fix_bf16(convert_from_huggingface(part1, model, 32, 8)), strict=False)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/nn/state.py", line 71, in load_state_dict
    v.assign(state_dict[k].shard(mlb.device, mlb.axis) if isinstance((mlb:=v.lazydata), MultiLazyBuffer) else state_dict[k].to(v.device)).realize()
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/tensor.py", line 133, in realize
    Tensor.corealize([self])
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/tensor.py", line 130, in corealize
    run_schedule(create_schedule(flatten([x.lazydata.lbs if isinstance(x.lazydata, MultiLazyBuffer) else [x.lazydata] for x in lst])))
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/realize.py", line 56, in run_schedule
    prg = lower_schedule_item(si)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/realize.py", line 47, in lower_schedule_item
    return Device[si.out.device].get_runner(si.ast)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/device.py", line 269, in get_runner
    def get_runner(self, ast:LazyOp) -> CompiledASTRunner: return self.to_program(self.get_linearizer(ast))
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/device.py", line 234, in to_program
    ret = CompiledASTRunner(k.ast, k.name, self.compiler.render(to_function_name(k.name), k.uops), self, k.global_size, k.local_size)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/device.py", line 196, in __init__
    lib:bytes = precompiled if precompiled is not None else self.device.compiler.compile_cached(prg)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/device.py", line 183, in compile_cached
    lib = self.compile(src)
  File "/Users/rasmus/Desktop/DEV/AI/tinygrad/tinygrad/runtime/ops_clang.py", line 15, in compile
    subprocess.check_output(args=('clang -shared -march=native -O2 -Wall -Werror -x c -fPIC - -o '+ \
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['clang', '-shared', '-march=native', '-O2', '-Wall', '-Werror', '-x', 'c', '-fPIC', '-', '-o', '/var/folders/x2/cvsn6mc905xdvtrw64wq0xcc0000gn/T/tmpf7zo7jvl']' returned non-zero exit status 1.
avg:     0.00 GFLOPS    13.38 GB/s           total:     2 kernels     0.00 GOPS     0.03 GB     2.51 ms
```

Edit:

on Macbook Pro w M2 Pro chip @ macOS Ventura 13.3

**Comment by geohot** [comment]
Both work for me on M3. Are you on master?

**Comment by skoshx** [comment]
Yeah, on master, freshly cloned from GitHub just before creating this issue

**Comment by skoshx** [comment]
@geohot 

Overriding `bfloat16` value to `half` in `dtype.py` fixes this. 

```python
bfloat16: Final[DType] = DType(10, 2, "half", None, 1)
```

So the problem seems to just be that the code uses `bfloat16` even if it's not supported on M2 according to [some sources](https://developer.apple.com/forums/thread/726201)

There is a comment `TODO: check if device supports bf16` in the `fix_bf16`, so maybe has to do with that

**Comment by geohot** [comment]
Can you make a failing test on the METAL CI?

**Comment by skoshx** [comment]
@geohot added a minimal test case... It has also the error "error: C-style cast from 'float' to '__bf16' is not allowed"

**Comment by nullhook** [comment]
@skoshx what you've linked to discusses CPU ISA and not the GPU. 

the proper way to solve this is to check if the device supports the following families: Metal3, Apple6, Mac2, as only these devices have bfloat hardware support, according to Metal's feature set table.

<img width="1313" alt="image" src="https://github.com/tinygrad/tinygrad/assets/8665427/db9fb53f-f8e2-42af-8160-b777dfb3d15c">

**Comment by skoshx** [comment]
@nullhook thanks for this! But doesn't everything that is run using `METAL` device run on the GPU? And isn't the test correct nonetheless?

**Comment by skoshx** [comment]
Closing this for now, since M2 Pro should support `bfloat16` according to the Metal docs. If someone else faces the same problem with the M2 Pro, please reply and I will reopen.

**Comment by marschr** [comment]
tl;dr: I had to install clang-18 instead of the shipped clang-14 on ubuntu 22.04 for examples/llama.py to work.

I'm on ubuntu 22.04 running inference on an old Xeon CPU, the clang 14 shipped with ubuntu 22.04 didn't seem to have `__bf16` support. After installing clang-18 and linking the binaries (i.e. `sudo ln -s /usr/lib/llvm-18/bin/clang /usr/bin/clang`) everything worked. Don't know how much older versions of mac OS/Xcode relies on old clang and may also affect Metal compilation.
I've only tried with clang-18 so there's a chance that clang 15~17 also works, needs testing. ( Also LLVM issue from aug/2022 maybe related: https://reviews.llvm.org/D130964 )

**Comment by skoshx** [comment]
Yeah, I actually do have Clang 14 on my machine. Tried with Clang 17 by changing `ops_clang.py` to:
```python
subprocess.check_output(args=('/opt/homebrew/Cellar/llvm/17.0.6_1/bin/clang -shared -march=native -O2 -Wall -Werror -x c -fPIC - -o '+ \
                                    str(output_file.name)).split(), input=(CLANG_PROGRAM_HEADER+src).encode('utf-8'))
```

And it worked!! Metal still doesn't work, so the problem seems to be that Metal uses Clang 14 which doesn't support bfloat16?

**Comment by nullhook** [comment]
the issue is simple: if the metal device isn't part of the above families, then it simply does not support hardware bfloats, and you need to fallback to using half or full precision types.

to solve this, you need a device feature flag in `ops_metal.py`, and in `cstyle.py`, you need visibility over the device's object. from there, it would be a conditional render.

**Comment by skoshx** [comment]
Yeah, but this seems to be also a problem for CLANG, and it would be nice to have some kind of helpful error message like "Please upgrade your Clang installation to version X to use `bfloat16`".

Anyways, as I was installing different versions of clang through Homebrew, somehow the error wen't completely away with clang. I cannot even reproduce this anymore. This feels super strange and doesn't make any sense that a Homebrew would affect the Apple CommandLineTools install.

The error does however still happen on Metal.

**Comment by chenyuxyz** [comment]
closing as stale, feel free to open new issues
